libsolidity/formal/BMC.cpp

```diff
@@ -184,7 +185,42 @@ bool BMC::visit(IfStatement const& _node)
 		);
 	m_context.popSolver();
 
-	SMTEncoder::visit(_node);
+	auto mergePathConditions = [](smtutil::Expression const& trueCondition, smtutil::Expression const& falseCondition) {
+		return trueCondition || falseCondition;
+	};
+
+	// from SMTEncoder::visit
+	_node.condition().accept(*this);
+	auto conditionExpr = expr(_node.condition());
+	// visit true branch
+	auto indicesBeforeBranch = copyVariableIndices();
+	pushPathCondition(conditionExpr);
+	_node.trueStatement().accept(*this);
+	auto trueEndPathCondition = currentPathConditions();
+	popPathCondition();
+	auto indicesEndTrue = copyVariableIndices();
+	resetVariableIndices(indicesBeforeBranch);
+	auto touchedVars = touchedVariables(_node.trueStatement());
+
+	decltype(indicesEndTrue) indicesEndFalse;
+	pushPathCondition(!conditionExpr);
+	auto falseEndPathCondition = currentPathConditions();
+	if (_node.falseStatement())
+	{
+		indicesBeforeBranch = copyVariableIndices();
+		_node.falseStatement()->accept(*this);
+		falseEndPathCondition = currentPathConditions();
+		indicesEndFalse = copyVariableIndices();
+		resetVariableIndices(indicesBeforeBranch);
+		touchedVars += touchedVariables(*_node.falseStatement());
+	}
+	else
+		indicesEndFalse = copyVariableIndices();
+
+	popPathCondition();
+
+	setPathCondition(mergePathConditions(trueEndPathCondition, falseEndPathCondition));
```

Comment:
Is the wanted behavior here really `set`? Wouldn't it be more precise to `push`?

Reply:
Actually I don't really get why you need to add this path condition.